src/hotspot/share/oops/objArrayKlass.cpp

```diff
@@ -312,7 +315,14 @@ Klass* ObjArrayKlass::array_klass(int n, TRAPS) {
 
   assert(dimension() <= n, "check order of chain");
   int dim = dimension();
-  if (dim == n) return this;
+  if (dim == n) {
+#if INCLUDE_CDS
+  if (UseSharedSpaces) {
+    DynamicArchive::log_array_class_load(THREAD, this);
+  }
+#endif
```

Comment:
The `ObjArrayKlass::array_klass(int n, TRAPS)` function can be called multiple times, so you'd have duplicated output of "class, load, cds" for the same array class.

I think it's better to move the logging code from `DynamicArchive::log_array_class_load()` into a function like `ArrayClass::log_array_class_load()`. It can be called from:

- `ArrayKlass(Symbol* name, KlassKind kind)` constructor
- `ArrayKlass::restore_unshareable_info()`

That way we can catch the actual loading event, and also print out whether the class is in static CDS archive, dynamic CDS archive, or not in CDS.

Instead of `(class, load, cds)`, maybe we can use a new tag like `(class,load,array)`?

Reply:
I've made the above change.
For array classes not loaded from CDS archive, there won't be `source` info.
Example output:
```
[0.048s][debug][class,load,cds] [[Ljava.lang.Object; source: shared objects file
[0.054s][debug][class,load,cds] [[[J source: shared objects file (top)
[0.065s][debug][class,load,cds] [Ljava.net.URL;
```